### PR TITLE
[ARM] `az bicep lint`: Add new command to lint a bicep file

### PR DESCRIPTION
**Pull request by polatengin**
**Related command**

`az bicep lint` command is added

**Description**
_"bicep cli"_ has a _lint_ command that lints the given _bicep_ file.

**History Notes**

[ARM] `az bicep lint`: Add new command to lint a bicep file

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ARM

**Comment by johnnyreilly** [comment]
Is there a release date for this version of the az CLI please?

**Comment by polatengin** [comment]
> Is there a release date for this version of the az CLI please?

in the `Milestone` section right panel of this page, I see [October 2023 - 2023-11-15](https://github.com/Azure/azure-cli/milestone/135)

**Comment by polatengin** [comment]
@yonzhan , @zhoxing-ms can I kindly ask for a review for this PR, please?

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 1]
```diff
@@ -3102,6 +3102,18 @@
...
+type: command
+short-summary: Lints a Bicep file.
+examples:
+  - name: Lints a Bicep file.
+    text: az bicep lint --file {bicep_file}
+  - name: Lints a Bicep file without restoring external modules.
+    text: az bicep lint --file {bicep_file} --no-restore
+  - name: Lints a Bicep file with specified diagnostics format. Valid values are ( default | sarif ).
+    text: az bicep lint --file {bicep_file} --diagnostics-format {diagnostics_format}
+"""
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:3115]: It is better to keep consistence with other help messages and use first-person verbs.
```suggestion
helps['bicep lint'] = """
type: command
short-summary: Lint a Bicep file.
examples:
  - name: Lint a Bicep file.
    text: az bicep lint --file {bicep_file}
  - name: Lint a Bicep file without restoring external modules.
    text: az bicep lint --file {bicep_file} --no-restore
  - name: Lint a Bicep file with specified diagnostics format. Valid values are ( default | sarif ).
    text: az bicep lint --file {bicep_file} --diagnostics-format {diagnostics_format}
"""
```

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 2]
```diff
@@ -863,6 +863,11 @@ def load_arguments(self, _):
         c.argument('output_format', help="Set output format. Valid values are ( json | bicepparam ).")
         c.argument('include_params', help="Set include params. Valid values are ( all | required-only ).")
 
+    with self.argument_context('bicep lint') as c:
+        c.argument('file', arg_type=bicep_file_type, help="The path to the Bicep module file to lint in the file system.")
+        c.argument('no_restore', arg_type=bicep_no_restore_type, help="When set, generates the parameters file without restoring external modules.")
+        c.argument('diagnostics_format', help="Set diagnostics format. Valid values are ( default | sarif ).")
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:869]: If the valid values for this param are enumerable, please add `arg_type=get_enum_type` to the param, and then CLI will automatically list the available values to customers and validate the input.
```suggestion
        c.argument('diagnostics_format', arg_type=get_enum_type(['default', 'sarif']), help="Set diagnostics format.")
```

**Comment by polatengin** [comment]
thanks for the comments @yanzhudd ! with the latest commits, they're all addressed 👍

**Comment by zhoxing-ms** [comment]
@Shenglol Could you please help review the bicep related commands?

**Review by shenglol — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 3]
```diff
@@ -4432,6 +4432,24 @@ def generate_params_file(cmd, file, no_restore=None, outdir=None, outfile=None,
         logger.error("az bicep generate-params could not be executed with the current version of Bicep CLI. Please upgrade Bicep CLI to v%s or later.", minimum_supported_version)
 
 
+def lint_bicep_file(cmd, file, no_restore=None, diagnostics_format=None):
+    ensure_bicep_installation(cmd.cli_ctx)
+
+    minimum_supported_version = "0.7.4"
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:4438]: I have seen many duplicate definitions `minimum_supported_version = "0.7.4"` at present. Could we abstract them into a uniformly configurable constant? This is convenient for future maintenance.

Comment (polatengin) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:4438]: Although I agree with your comment, I don't think this PR should have changes for other places in code. I'm going to create a follow up PR to address this request, if that's OK.

Can I ask to finish review with my comment in mind, please?

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:4438]: OK, looks good to me

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**